### PR TITLE
Add CPU drawers and books option support for s390x

### PR DESCRIPTION
**Pull request by fbq815**
After this patch, we can define a 5-level CPU hierarchy on s390x like:
cpus=,drawers=,books=,sockets=,cores=,maxcpus=

ID: 2036

Signed-off-by: Boqiao Fu <bfu@redhat.com>

### CODE REVIEW

**Comment by fbq815** [comment]
@smitterl @PaulYuuu @YongxueHong @zhenyzha @maxujun @nanliu-r please help me review this patch

**Comment by smitterl** [comment]
@fbq815 Are there some tests that already use these new values? Which ones? Can you paste test results here?

**Comment by fbq815** [comment]
> @fbq815 Are there some tests that already use these new values? Which ones? Can you paste test results here?

not yet, I'll add some after this is merged

**Comment by smitterl** [comment]
> > @fbq815 Are there some tests that already use these new values? Which ones? Can you paste test results here?
> 
> not yet, I'll add some after this is merged

Got it. On avocado-vt/tp-libvirt we're requested to always show passing tests to make sure the new code works as expected. I don't know if that's different for avocado-vt/tp-qemu. I would assume yes. So if you add a new patch in tp-qemu you would reference it in your description or some comment like "covered via <ref to PR>", e.g. like here: https://github.com/avocado-framework/avocado-vt/pull/3840#issue-2111855964

**Comment by PaulYuuu** [comment]
Refer to this: https://gitlab.com/qemu-project/qemu/-/commit/5de1aff2555275ef182197eddcadb276364ace38 for the perfect compatibility and format.

**Comment by fbq815** [comment]
> > > @fbq815 Are there some tests that already use these new values? Which ones? Can you paste test results here?
> > 
> > 
> > not yet, I'll add some after this is merged
> 
> Got it. On avocado-vt/tp-libvirt we're requested to always show passing tests to make sure the new code works as expected. I don't know if that's different for avocado-vt/tp-qemu. I would assume yes. So if you add a new patch in tp-qemu you would reference it in your description or some comment like "covered via ", e.g. like here: [#3840 (comment)](https://github.com/avocado-framework/avocado-vt/pull/3840#issue-2111855964)

ok, I'll add related test case in tp-qemu then

**Comment by fbq815** [comment]
@smitterl I added the related test cases to this MR in tp-qemu, please help me check

**Comment by fbq815** [comment]
you could see books and draws calculate by this patch if you check the log of cpu_topology_test.check_core_per_sockets

Test result on s390x:
cpu_topology_test.siblings_core.s390-virtio: PASS (21.80 s)
cpu_topology_test.check_core_per_sockets.two_cores_in_one_socket.s390-virtio: PASS (21.56 s)
cpu_topology_test.check_core_per_sockets.one_cores_per_socket.s390-virtio: PASS (21.61 s)
cpu_topology_test.check_core_per_sockets.two_cores_per_socket.s390-virtio: PASS (22.12 s)

File: virttest/qemu_vm.py
Comment (smitterl) [inline virttest/qemu_vm.py:1965]: Could these conditionals about vcpu drawers and books be removed when setting the default value at creation?
```
        vcpu_drawers = int(params.get("vcpu_drawers", 1))
        enable_drawers = vcpu_drawers != "INVALID" and Flags.SMP_DRAWERS in devices.caps
```

**Review by smitterl — APPROVED** [review]
Thank you, I have a question but rest LGTM.

File: virttest/qemu_vm.py
Comment (fbq815) [reply virttest/qemu_vm.py:1965]: yes, vt will give the default value of 1

**Comment by fbq815** [comment]
@PaulYuuu could you please help me merge this MR

**Review by PaulYuuu — APPROVED** [review]

**Comment by fbq815** [comment]
@YongxueHong could you please help me merge this patch?

**Comment by nanliu-r** [comment]
`cat /proc/cpuinfo |grep siblings |uniq |cut -d ":" -f 2`  Can get the same results on rhel8 and rhel9 host on x86.

**Review by YongxueHong — APPROVED** [review]
LGTM.

**Review by zhenyzha — APPROVED** [review]
Test Pass on aarch64

LGTM.Acked-by: Zhenyu Zhang [zhenyzha@redhat.com](mailto:zhenyzha@redhat.com)

**Pull request merged**
